src/bookshelf/src/bksw_sec.erl

```diff
@@ -35,109 +37,283 @@ is_authorized(Req0, #context{} = Context) ->
+%-spec get_signed_headers(proplist(), proplist(), proplist()) -> proplist(). % for erlang20+
+-spec get_signed_headers(SignedHeaderKeys::[string()], Headers::[tuple()], SignedHeaders::[tuple()]) -> [tuple()].
+get_signed_headers([], _, SignedHeaders) -> lists:reverse(SignedHeaders);
+get_signed_headers(_, [], SignedHeaders) -> lists:reverse(SignedHeaders);
+get_signed_headers([Key | SignedHeaderKeys], Headers0, SignedHeaders) ->
+    {_, SignedHeader, Headers} = lists:keytake(Key, 1, Headers0),
+    get_signed_headers(SignedHeaderKeys, Headers, [SignedHeader | SignedHeaders]).
+
+% split authorization header into component parts
+% https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+- spec parse_authorization(string()) -> [string()] | err.
+parse_authorization(Auth) ->
+    case string:split(Auth, " ", all) of
+        ["AWS4-HMAC-SHA256", "Credential="++Cred, "SignedHeaders="++SigHead, "Signature="++Signature] ->
+            [string:trim(Cred, trailing, ","), string:trim(SigHead, trailing, ","), Signature];
         _ ->
-            encode_sign_error_response(AccessKeyId, Signature,
-                                       RequestId, StringToSign, Req0, Context)
+            err
     end.
 
--spec expire_diff(undefined | binary()) -> integer().
-expire_diff(undefined) -> 1;
-expire_diff(Expires) ->
-    Now = calendar:datetime_to_gregorian_seconds(erlang:universaltime()),
-    bksw_util:to_integer(Expires) - (Now - ?SECONDS_AT_EPOCH).
+% split credentials string into component parts
+% Cred = "<access-key-id>/<date>/<AWS-region>/<AWS-service>/aws4_request"
+- spec parse_x_amz_credential(string()) -> [string()].
+parse_x_amz_credential(Cred) ->
+    Parse = string:split(Cred, "/", all),
+    case Parse of
+        [_access_key_id, _date, _aws_region, "s3", "aws4_request"] -> Parse;
+        _                                                          -> err
+    end.
+
+% split signed header string into component parts. returns empty string on empty string.
+% Headers = "<header1>;<header2>;...<headerN>"
+- spec parse_x_amz_signed_headers(string()) -> [string()].
+parse_x_amz_signed_headers(Headers) ->
+   string:split(Headers, ";", all).
+
+% convert the keys of key-value pairs to lowercase strings
+%-spec process_headers([proplist()]) -> [proplist()]. % for erlang20+
+-spec process_headers(Headers::[tuple()]) -> [tuple()].
+process_headers(Headers) ->
+    [{string:casefold(
+        case is_atom(Key) of
+            true -> atom_to_list(Key);
+            _    -> Key
+        end), Val} || {Key, Val} <- Headers].
 
-%% get_bucket([Bucket, _, _]) ->
-%%     Bucket.
+-spec is_expired(DateTimeString::string(), ExpiresSec::integer()) -> boolean().
+is_expired(DateTimeString, ExpiresSec) ->
+    % most ways of getting the date/time seem problematic.  for instance, docs for
+    % calendar:universal_time() and erlang:universaltime() say: 'Returns local time
+    % if universal time is unavailable.'
```

Comment:
This comment is a true nerd-snipe, beware.  You may ask.  "Wait, under what conditions will this happen?"  What you will quickly find is that calendar:universal_time and erlang:universaltime() are the same function (the former calls the latter).  erlang:universaltime is implemented in C. It calls `time` followed by `gmtime_r`.  `gmtime_r` does the conversion into universal time.  From there the code depends on which libc we are linked against.  In glibc, we will eventually call`__tz_convert` with `use_localtime` set to 0:

https://sourceware.org/git/?p=glibc.git;a=blob;f=time/tzset.c;h=16b20b96d3092fae969eebff0c2aaeaf9a9012ca;hb=HEAD#l566

At this point I began to regret my life choices and have decided to let this be for now.

Reply:
I'm impressed that you went through all the trouble of finding that out!